src/Microsoft.Extensions.DependencyInjection/ServiceLookup/CallSiteExpressionBuilder.cs

```diff
@@ -0,0 +1,233 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Threading;
+
+namespace Microsoft.Extensions.DependencyInjection.ServiceLookup
+{
+    internal class CallSiteExpressionBuilder: CallSiteVisitor<ParameterExpression, Expression>
+    {
+        private static readonly MethodInfo CaptureDisposableMethodInfo = GetMethodInfo<Func<ServiceProvider, object, object>>((a, b) => a.CaptureDisposable(b));
+        private static readonly MethodInfo TryGetValueMethodInfo = GetMethodInfo<Func<IDictionary<object, object>, object, object, bool>>((a, b, c) => a.TryGetValue(b, out c));
+        private static readonly MethodInfo AddMethodInfo = GetMethodInfo<Action<IDictionary<object, object>, object, object>>((a, b, c) => a.Add(b, c));
+        private static readonly MethodInfo MonitorEnterMethodInfo = GetMethodInfo<Action<object, bool>>((lockObj, lockTaken) => Monitor.Enter(lockObj, ref lockTaken));
+        private static readonly MethodInfo MonitorExitMethodInfo = GetMethodInfo<Action<object>>(lockObj => Monitor.Exit(lockObj));
+
+        private readonly ParameterExpression _providerParameter = Expression.Parameter(typeof(ServiceProvider));
+
+        private readonly List<Expression> _initializations = new List<Expression>();
+        private readonly List<ParameterExpression> _variables = new List<ParameterExpression>();
+        private readonly List<Expression> _locks = new List<Expression>();
+
+        private readonly Dictionary<Expression, ParameterExpression> _resolvedServices = new Dictionary<Expression, ParameterExpression>();
+        private readonly Dictionary<Expression, LambdaExpression> _captureDisposable = new Dictionary<Expression, LambdaExpression>();
+
+        private ParameterExpression _rootVariable;
+
+        public Expression<Func<ServiceProvider, object>> Build(IServiceCallSite callSite)
+        {
+            var serviceExpression = VisitCallSite(callSite, _providerParameter);
+
+            List<Expression> body = new List<Expression>();
+            body.AddRange(_initializations);
+            foreach (var expression in _locks)
+            {
+                serviceExpression = Lock(serviceExpression, expression);
+            }
+            body.Add(serviceExpression);
+
+            return Expression.Lambda<Func<ServiceProvider, object>>(
+                Expression.Block(_variables, body),
+                _providerParameter);
+        }
+
+        protected override Expression VisitSingleton(SingletonCallSite singletonCallSite, ParameterExpression provider)
+        {
+            return VisitScoped(singletonCallSite, GetRootProvider());
```

Comment:
I'm no longer sold on building complicated expression trees for singletons, since it always winds up being one dictionary lookup once the tree is built in practice anyways. Wouldn't it be better to avoid CAS altogether for the root `_resolvedServices` by creating an expression that calls `CallSiteRuntimeResolver.Resolve(singletonCallSite, provider);` like #426.